### PR TITLE
[ASCII-2614] Add a task to generate a dependency graph

### PR DESCRIPTION
**Pull request by pgimalac**
### What does this PR do?
Add an invoke task to generate a dependency graph using https://github.com/loov/goda.

### Motivation
We have a number of build tags on our builds, and different build tags per process/platform, which makes using Goda way more complex. 
This invoke tasks make it trivial to generate dependency graphs.

### Describe how you validated your changes

Tested manually.

### CODE REVIEW

File: tasks/go_deps.py [hunk 1]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+          inv -e go-deps.graph --build process-agent --os linux --target k8s.io/...
+        Dependency graph of workloadmeta on Linux when using the same build tags as the core-agent
+          inv -e go-deps.graph --build agent --os linux --entrypoint github.com/DataDog/datadog-agent/comp/core/workloadmeta/...:all
+    """
+    if os is None:
+        goos = ctx.run("go env GOOS", hide=True)
+        assert goos
+        os = goos.stdout.strip()
+    else:
+        assert os in GOOS_MAPPING.values()
```
Comment (louis-cqrl) [inline tasks/go_deps.py:235]:
```suggestion
    os = os or ctx.run("go env GOOS", hide=True).stdout.strip()
```
same for `arch` value, old memory from doing some python, may need to be tested

File: tasks/go_deps.py [hunk 2]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+        arch = goarch.stdout.strip()
+    else:
+        assert arch in GOARCH_MAPPING.values()
+
+    stdarg = "-std" if std else ""
+    clusterarg = "-cluster" if cluster else ""
+
+    if entrypoint is None:
+        entrypoint = BINARIES[build]["entrypoint"]
+        entrypoint = f"github.com/DataDog/datadog-agent/{entrypoint}:all"
```
Comment (louis-cqrl) [inline tasks/go_deps.py:249]:
```suggestion
    entrypoint = entrypoint or f"github.com/DataDog/datadog-agent/{BINARIES[build]['entrypoint']}:all"
```

File: tasks/go_deps.py [hunk 3]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+
+    build_tags = get_default_build_tags(
+        build=build, flavor=AgentFlavor[flavor], platform=key_for_value(GOOS_MAPPING, os)
+    )
+    for tag in build_tags:
+        entrypoint = f"{tag}=1({entrypoint})"
+
+    expr = entrypoint
+    if target is not None:
+        expr = f"reach({entrypoint}, {target})"
```
Comment (louis-cqrl) [inline tasks/go_deps.py:259]:
```suggestion
    expr = entrypoint if target is None else f"reach({entrypoint}, {target})"
```

File: tasks/go_deps.py [hunk 4]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+        os = goos.stdout.strip()
+    else:
+        assert os in GOOS_MAPPING.values()
+
+    if arch is None:
+        goarch = ctx.run("go env GOARCH", hide=True)
+        assert goarch
+        arch = goarch.stdout.strip()
+    else:
+        assert arch in GOARCH_MAPPING.values()
```
Comment (louis-cqrl) [inline tasks/go_deps.py:242]:
```suggestion
    arch = arch or ctx.run("go env GOARCH", hide=True).stdout.strip()
```

**Review by louis-cqrl — APPROVED** [review]
left small suggestions, some may need to be tested, I don't know how often this code will be used, or changed so feel free to ignore them.

File: tasks/go_deps.py (hunk 1 above)
Comment (louis-cqrl) [reply tasks/go_deps.py:235]: You can add the assertion just below, I forget them

File: tasks/go_deps.py [hunk 5]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+        Dependency graph of the trace-agent on Linux
+          inv -e go-deps.graph --build trace-agent --os linux
+        Reachability graph of the process-agent on Linux to k8s.io/... dependencies
+          inv -e go-deps.graph --build process-agent --os linux --target k8s.io/...
+        Dependency graph of workloadmeta on Linux when using the same build tags as the core-agent
+          inv -e go-deps.graph --build agent --os linux --entrypoint github.com/DataDog/datadog-agent/comp/core/workloadmeta/...:all
+    """
+    if os is None:
+        goos = ctx.run("go env GOOS", hide=True)
+        assert goos
```
Comment (ogaca-dd) [inline tasks/go_deps.py:232]: Note: As far as I know, the idiomatic way to report an error when using invoke is to use `raise Exit()`

File: tasks/go_deps.py (hunk 1 above)
Comment (ogaca-dd) [inline tasks/go_deps.py:235]: Would it make sense to check if os is in GOOS_MAPPING also when it is provided from `go env`?

File: tasks/go_deps.py [hunk 6]
```diff
@@ -185,3 +186,95 @@ def show(ctx: Context, build: str, flavor: str = AgentFlavor.base.name, os: str
...
+    entrypoint: str | None = None,
+    target: str | None = None,
+    std: bool = False,
+    cluster: bool = False,
+    fmt: str = "svg",
+    auto_open: bool = True,
+):
+    """
+    Generate a dependency graph of the given build.
+    Requires https://github.com/loov/goda and Graphviz's `dot` tools to be in the PATH.
```
Comment (ogaca-dd) [inline tasks/go_deps.py:220]: What do you think about checking if `goda` / `dot` are in the `PATH` and provides a link if needed? Alternatively, you can install these tools in the invoke command `inv setup`.

**Review by ogaca-dd — APPROVED** [review]
Very useful invoke task!

File: tasks/go_deps.py (hunk 2 above)
Comment (pgimalac) [reply tasks/go_deps.py:249]: I think the current way is more readable

File: tasks/go_deps.py (hunk 4 above)
Comment (pgimalac) [reply tasks/go_deps.py:242]: I agree but to please the type system I need to assert the return value of `ctx.run` is not `None`

File: tasks/go_deps.py (hunk 1 above)
Comment (pgimalac) [reply tasks/go_deps.py:235]: I agree but to please the type system I need to assert the return value of `ctx.run` is not `None`

Comment (pgimalac) [reply tasks/go_deps.py:235]: Yes good catch I'll update that !

File: tasks/go_deps.py (hunk 5 above)
Comment (pgimalac) [reply tasks/go_deps.py:232]: You are right, but as this is not a "normal" exit path but just a code assertion, `ctx.run` is not supposed to return `None`, so I think it's fine to check that way (also it's shorter and simpler than having an `if` block, and we would have a stack trace to help with debugging)

File: tasks/go_deps.py (hunk 6 above)
Comment (pgimalac) [reply tasks/go_deps.py:220]: Yes I agree having a clear error message would be much better, good idea !
I think most people will never use this task (it's very niche), it's probably not useful to install it as part of `inv setup`.

**Pull request merged**